x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/PrecompiledCharMapNormalizer.java

```diff
@@ -155,40 +155,47 @@ String normalize(String str) {
         // break points from there. But, this seemed the easiest way for now
         //
         // Keep in mind, these break points aren't necessarily surrogate pairs, but also codepoints that contain a combining mark
+        byte[] strBytes = str.getBytes(StandardCharsets.UTF_8);
+        char[] strChars = str.toCharArray();
+        int[] strCp = str.codePoints().toArray();
         BreakIterator b = BreakIterator.getCharacterInstance(Locale.ROOT);
         b.setText(str);
-        int start = b.first();
-        // If we knew the utf-8 length ahead of time (and iterated over the bytes in the appropriate chunks)
-        // we could pre-populate the known length here.
+        // We iterate the whole string, so b.first() is always `0`
+        int startIter = b.first();
+        int codePointPos = 0;
         BytesRefBuilder strBuilder = new BytesRefBuilder();
-        for (int end = b.next(); end != BreakIterator.DONE; start = end, end = b.next()) {
-            // TODO: It would be awesome if we could translate these starts and ends to byte positions, if we could performance would be
-            // dramatically improved
-            String unicodeStr = str.substring(start, end);
-            byte[] unicode = unicodeStr.getBytes(StandardCharsets.UTF_8);
+        strBuilder.grow(strBytes.length);
+        int bytePos = 0;
+        for (int end = b.next(); end != BreakIterator.DONE; startIter = end, end = b.next()) {
+            int byteLen = 0;
+            int numCp = str.codePointCount(startIter, end);
+            for (int i = codePointPos; i < numCp + codePointPos; i++) {
+                byteLen += numUtf8Bytes(strCp[i]);
+            }
+            codePointPos += numCp;
```

Comment:
@droberts195 we always start at `0`, but here we only iterate codepoints based on the previously counted code points. So it should satisfy your concerns.